# Tictactrip - Technical test 

By Vincent DEBANDE

In [1]:
import pandas as pd 
import seaborn as sns
import numpy as np

## 1. Loading data

Let's load our data !

### 1.1 Tickets data

Here are tickets from tictactrip (one row = one ticket) :

- id
- company that assure the journey
- o_station : Station of origin 
- d_station : Station of destination
- departure_ts : departure time
- arrival_ts : arrival_time
- price_in_cents
- search_ts : ? 
- middle_stations : if there are any, list of the stops
- other_companies : a journey can be assured by multiples companies
- o_city : city of origin
- d_city : city of destination

In [2]:
tickets = pd.read_csv("./data/ticket_data.csv")
tickets.fillna(-1, inplace = True)

tickets['o_station'] = tickets['o_station'].astype('int64')
tickets['d_station'] = tickets['d_station'].astype('int64')

tickets

id  company  o_station  d_station            departure_ts  \
0      6795025     8385         -1         -1  2017-10-13 14:00:00+00   
1      6795026        9         63       1044  2017-10-13 13:05:00+00   
2      6795027     8377       5905       6495  2017-10-13 13:27:00+00   
3      6795028     8377       5905       6495  2017-10-13 13:27:00+00   
4      6795029     8381       5905       6495  2017-10-13 21:46:00+00   
...        ...      ...        ...        ...                     ...   
74163  6869182       13        279      10729  2017-10-27 02:30:00+00   
74164  6869185        9        279        304  2017-10-27 07:00:00+00   
74165  6869187     8371      10642        304  2017-10-27 08:30:00+00   
74166  6869189       13        279        304  2017-10-27 13:25:00+00   
74167  6869191     8371      10642      10729  2017-10-27 14:45:00+00   

                   arrival_ts  price_in_cents                   search_ts  \
0      2017-10-13 20:10:00+00            4550  2017-10-01 00:13:31.327+00   
1      2017-10-14 06:55:00+00            1450  2017-10-01 00:13:35.773+00   
2      2017-10-14 21:24:00+00            7400  2017-10-01 00:13:40.212+00   
3      2017-10-14 11:02:00+00           13500  2017-10-01 00:13:40.213+00   
4      2017-10-14 19:32:00+00            7710  2017-10-01 00:13:40.213+00   
...                       ...             ...                         ...   
74163  2017-10-27 16:00:00+00            2590  2017-10-19 10:35:42.943+00   
74164  2017-10-27 13:30:00+00            2700  2017-10-19 10:35:42.943+00   
74165  2017-10-27 15:15:00+00            3000  2017-10-19 10:35:42.943+00   
74166  2017-10-27 21:10:00+00            2880  2017-10-19 10:35:42.943+00   
74167  2017-10-27 21:40:00+00            2000  2017-10-19 10:35:42.943+00   

           middle_stations other_companies  o_city  d_city  
0                       -1              -1     611     542  
1                {149,418}            {13}     611     542  
2      {798,798,6794,6246}     {8377,8376}     611     542  
3      {798,798,6794,6246}     {8377,8376}     611     542  
4              {5983,5983}          {8380}     611     542  
...                    ...             ...     ...     ...  
74163            {408,408}          {8371}     628     562  
74164          {1105,1105}             {9}     628     562  
74165          {1105,1105}             {9}     628     562  
74166            {863,863}            {13}     628     562  
74167            {863,863}          {8371}     628     562  

[74168 rows x 12 columns]

In [3]:
tickets.dtypes

id                  int64
company             int64
o_station           int64
d_station           int64
departure_ts       object
arrival_ts         object
price_in_cents      int64
search_ts          object
middle_stations    object
other_companies    object
o_city              int64
d_city              int64
dtype: object

Since there are some NaN in o_station and d_station, the type is float ! We'll deal with that later if needed.

### 1.2 Cities

Here are all the cities (o_city, d_city in tickets) : 

- id
- local_name : Name of the city, region, country
- unique_name : name of the city (unique) 
- coordinates (latitude / longitude)
- population

In [4]:
cities = pd.read_csv("./data/cities.csv")
cities

id                                      local_name  \
0      5159                           Padua, Veneto, Italia   
1        76                     Barcelona, Cataluña, España   
2        81                     Basel, Basel-Stadt, Schweiz   
3       259                   Erlangen, Bayern, Deutschland   
4     11979                              Balș, Olt, România   
...     ...                                             ...   
8035   9964  Saint-André-lez-Lille, Hauts-de-France, France   
8036  10863             Kiev, Шевченківський район, Ukraine   
8037  11144                           Brighton, England, UK   
8038  11377           Manises, Comunitat Valenciana, España   
8039  11537               South Cambridgeshire, England, UK   

                unique_name   latitude  longitude  population  
0                     padua  45.406435  11.876761    209678.0  
1                 barcelona  41.385064   2.173404   1611822.0  
2                     basel  47.593437   7.619812         NaN  
3                  erlangen  49.589674  11.011961    105412.0  
4                      balș  44.353354  24.095672         NaN  
...                     ...        ...        ...         ...  
8035  saint-andre-lez-lille  50.654250   3.047690         NaN  
8036                   kiev  50.450100  30.523400         NaN  
8037               brighton  50.822000  -0.137400         NaN  
8038                manises  39.493000  -0.462100         NaN  
8039   south-cambridgeshire  52.179600  -0.003400         NaN  

[8040 rows x 6 columns]

In [5]:
cities.dtypes

id               int64
local_name      object
unique_name     object
latitude       float64
longitude      float64
population     float64
dtype: object

### 1.3 Stations

Here all the stations :

- id : unique number to define a station
- unique_name
- coordinates (latitude / longitude)

In [6]:
stations = pd.read_csv("./data/stations.csv")
stations

id                                        unique_name   latitude  \
0          1                         Aalen (Stuttgarter Straße)  48.835296   
1          2                         Aéroport Bordeaux-Mérignac  44.830226   
2          3                                      Aéroport CDG   49.009900   
3          4                     Aéroport de Berlin-Schönefeld   52.389446   
4          5                                Aéroport de Dresden  51.123604   
...      ...                                                ...        ...   
11030  11032                          Porto Avda De Los Aliados  41.146922   
11031  11033  Rom Ciampino Flughafen Piazzale Leonardo da Vinci  41.804102   
11032  11034                                              Sants  41.379892   
11033  11035                                       Vozilici D66  45.157136   
11034  11036                     Wien Hotel ibis Vienna Airport  48.123229   

       longitude  
0      10.092956  
1      -0.700883  
2       2.559310  
3      13.520345  
4      13.764737  
...          ...  
11030  -8.611122  
11031  12.597110  
11032   2.139462  
11033  14.164140  
11034  16.437528  

[11035 rows x 4 columns]

In [7]:
stations.dtypes

id               int64
unique_name     object
latitude       float64
longitude      float64
dtype: object

### 1.4 Providers



In [8]:
providers = pd.read_csv("./data/providers.csv")
providers

id  company_id provider_id           name        fullname has_wifi  \
0       9           1         NaN         ouibus          Ouibus     True   
1      10           2         NaN        deinbus      Deinbus.de    False   
2      11           3         NaN        infobus         Infobus    False   
3      12           4         NaN  studentAgency  Student Agency    False   
4      13           5         NaN        flixbus         Flixbus     True   
..    ...         ...         ...            ...             ...      ...   
222  8384          10          16         thalys          Thalys     True   
223  8385          11         NaN            bbc       Blablacar    False   
224  8387          30         NaN          vatry           Vatry      NaN   
225  8388          40         NaN        beauval         Beauval      NaN   
226  8389          12         NaN        idvroom         idVroom    False   

    has_plug has_adjustable_seats has_bicycle transport_type  
0       True                 True       False            bus  
1      False                False       False            bus  
2      False                False       False            bus  
3      False                False       False            bus  
4      False                False       False            bus  
..       ...                  ...         ...            ...  
222     True                False        True          train  
223    False                False       False     carpooling  
224      NaN                  NaN         NaN            bus  
225      NaN                  NaN         NaN            bus  
226    False                False       False     carpooling  

[227 rows x 10 columns]

In [9]:
providers.dtypes

id                       int64
company_id               int64
provider_id             object
name                    object
fullname                object
has_wifi                object
has_plug                object
has_adjustable_seats    object
has_bicycle             object
transport_type          object
dtype: object

## 2. The mission 

I'm asked to extract several informations about the trajects : 

- Minimum/maximum/average price
- Minimum/maximum/average duration

- Difference of price and duration between transport type and length of the journey

### 2.1 Trajects 

Let's define a traject as following : 

A traject is a journey from a city of origin (o_city) to a city of destination (d_city)

In [10]:
trajects = tickets.groupby(['o_city','d_city']).count()
trajects

id  company  o_station  d_station  departure_ts  arrival_ts  \
o_city d_city                                                                 
5      23        5        5          5          5             5           5   
6      227       2        2          2          2             2           2   
       504      30       30         30         30            30          30   
       628       4        4          4          4             4           4   
       845     201      201        201        201           201         201   
...            ...      ...        ...        ...           ...         ...   
11938  126      72       72         72         72            72          72   
12124  1064      2        2          2          2             2           2   
12166  857       1        1          1          1             1           1   
12190  639      18       18         18         18            18          18   
       8937      1        1          1          1             1           1   

               price_in_cents  search_ts  middle_stations  other_companies  
o_city d_city                                                               
5      23                   5          5                5                5  
6      227                  2          2                2                2  
       504                 30         30               30               30  
       628                  4          4                4                4  
       845                201        201              201              201  
...                       ...        ...              ...              ...  
11938  126                 72         72               72               72  
12124  1064                 2          2                2                2  
12166  857                  1          1                1                1  
12190  639                 18         18               18               18  
       8937                 1          1                1                1  

[1437 rows x 10 columns]

We can see that we have 1437 different trajects ! 

### 2.2 Analysis of trajects

#### 2.2.1 Prices

In order to get statistics on the trajets we can group by all the tickets by **o_city** and **d_city**. 
Then, we can apply functions on the grouped lines. Since we want to compute multiple statistics for each group, we can user .agg() function !

In [11]:
trajects = tickets.groupby(['o_city','d_city'])[['price_in_cents']].agg(['min','max','mean', 'count'])
trajects

price_in_cents                           
                         min    max          mean count
o_city d_city                                          
5      23              18600  22000  20320.000000     5
6      227              9860  13650  11755.000000     2
       504              2000   8920   4042.666667    30
       628              2600   3190   2797.500000     4
       845               700   2420    864.626866   201
...                      ...    ...           ...   ...
11938  126              3000   5350   4204.861111    72
12124  1064             6000   7900   6950.000000     2
12166  857              5300   5300   5300.000000     1
12190  639               600    850    688.888889    18
       8937             9800   9800   9800.000000     1

[1437 rows x 4 columns]

#### 2.2.2 Time

We want for each traject some statiscts about the duration ! 

First, we need to convert the string into a datetime, let's use pd.to_datetime().
Then, we'll do a **group by** and an **aggregation** !

In [12]:
tickets['departure_ts'] = pd.to_datetime(tickets['departure_ts'])
tickets['arrival_ts'] = pd.to_datetime(tickets['arrival_ts'])

tickets.dtypes

id                               int64
company                          int64
o_station                        int64
d_station                        int64
departure_ts       datetime64[ns, UTC]
arrival_ts         datetime64[ns, UTC]
price_in_cents                   int64
search_ts                       object
middle_stations                 object
other_companies                 object
o_city                           int64
d_city                           int64
dtype: object

In [13]:
# Let's compute the travel time
tickets['travel_ts'] = tickets['arrival_ts'] - tickets['departure_ts']
tickets

id  company  o_station  d_station              departure_ts  \
0      6795025     8385         -1         -1 2017-10-13 14:00:00+00:00   
1      6795026        9         63       1044 2017-10-13 13:05:00+00:00   
2      6795027     8377       5905       6495 2017-10-13 13:27:00+00:00   
3      6795028     8377       5905       6495 2017-10-13 13:27:00+00:00   
4      6795029     8381       5905       6495 2017-10-13 21:46:00+00:00   
...        ...      ...        ...        ...                       ...   
74163  6869182       13        279      10729 2017-10-27 02:30:00+00:00   
74164  6869185        9        279        304 2017-10-27 07:00:00+00:00   
74165  6869187     8371      10642        304 2017-10-27 08:30:00+00:00   
74166  6869189       13        279        304 2017-10-27 13:25:00+00:00   
74167  6869191     8371      10642      10729 2017-10-27 14:45:00+00:00   

                     arrival_ts  price_in_cents                   search_ts  \
0     2017-10-13 20:10:00+00:00            4550  2017-10-01 00:13:31.327+00   
1     2017-10-14 06:55:00+00:00            1450  2017-10-01 00:13:35.773+00   
2     2017-10-14 21:24:00+00:00            7400  2017-10-01 00:13:40.212+00   
3     2017-10-14 11:02:00+00:00           13500  2017-10-01 00:13:40.213+00   
4     2017-10-14 19:32:00+00:00            7710  2017-10-01 00:13:40.213+00   
...                         ...             ...                         ...   
74163 2017-10-27 16:00:00+00:00            2590  2017-10-19 10:35:42.943+00   
74164 2017-10-27 13:30:00+00:00            2700  2017-10-19 10:35:42.943+00   
74165 2017-10-27 15:15:00+00:00            3000  2017-10-19 10:35:42.943+00   
74166 2017-10-27 21:10:00+00:00            2880  2017-10-19 10:35:42.943+00   
74167 2017-10-27 21:40:00+00:00            2000  2017-10-19 10:35:42.943+00   

           middle_stations other_companies  o_city  d_city       travel_ts  
0                       -1              -1     611     542 0 days 06:10:00  
1                {149,418}            {13}     611     542 0 days 17:50:00  
2      {798,798,6794,6246}     {8377,8376}     611     542 1 days 07:57:00  
3      {798,798,6794,6246}     {8377,8376}     611     542 0 days 21:35:00  
4              {5983,5983}          {8380}     611     542 0 days 21:46:00  
...                    ...             ...     ...     ...             ...  
74163            {408,408}          {8371}     628     562 0 days 13:30:00  
74164          {1105,1105}             {9}     628     562 0 days 06:30:00  
74165          {1105,1105}             {9}     628     562 0 days 06:45:00  
74166            {863,863}            {13}     628     562 0 days 07:45:00  
74167            {863,863}          {8371}     628     562 0 days 06:55:00  

[74168 rows x 13 columns]

In [14]:
# We check the types of the new columns
tickets.dtypes

id                               int64
company                          int64
o_station                        int64
d_station                        int64
departure_ts       datetime64[ns, UTC]
arrival_ts         datetime64[ns, UTC]
price_in_cents                   int64
search_ts                       object
middle_stations                 object
other_companies                 object
o_city                           int64
d_city                           int64
travel_ts              timedelta64[ns]
dtype: object

In [15]:
trajects_Time = tickets.groupby(['o_city','d_city'])[['travel_ts']].agg(['min','max', 'sum', 'count'])
trajects_Time = trajects_Time['travel_ts']

# We round the timedelta to the second !
trajects_Time['mean'] = (trajects_Time['sum']/trajects_Time['count']).round('s')

# Since we already computed the mean, we can drop the sum column
trajects_Time = trajects_Time.drop('sum', 1)

trajects_Time

min             max  count            mean
o_city d_city                                                       
5      23     0 days 08:53:00 0 days 15:54:00      5 0 days 10:18:48
6      227    0 days 12:24:00 0 days 15:01:00      2 0 days 13:42:30
       504    0 days 05:36:00 0 days 12:20:00     30 0 days 08:17:24
       628    0 days 09:40:00 0 days 14:30:00      4 0 days 12:10:00
       845    0 days 01:00:00 0 days 04:11:00    201 0 days 01:19:55
...                       ...             ...    ...             ...
11938  126    0 days 05:30:00 1 days 11:20:00     72 0 days 07:25:17
12124  1064   0 days 11:10:00 0 days 23:10:00      2 0 days 17:10:00
12166  857    0 days 21:55:00 0 days 21:55:00      1 0 days 21:55:00
12190  639    0 days 01:10:00 0 days 02:40:00     18 0 days 01:28:53
       8937   0 days 09:10:00 0 days 09:10:00      1 0 days 09:10:00

[1437 rows x 4 columns]

### 2.3 Comparison between transport types based on length of the journey

Now we have to compute the deltas of the prices et duration based on the transport type and the number of km of the trajects. 

#### Transport type

To get the transport type use during a traject, we have to link tickets with providers. 

For tickets with multiple companies, we'll check if there are multiple types of transports. If there are, we'll ignore the ticket because we don't have the details of the journey (price and duration per sub journey).

In [16]:
tickets2 = tickets.copy()

In [17]:
def getTransportType(company, other_companies):
    if(other_companies == -1):
        return providers[providers['id'] == company]['transport_type']
    else:
        other_companies = set(map(int, other_companies[1:-1].split(',')))
        transport = providers[providers['id'] == company][['transport_type']].iloc[0]['transport_type']
        for company in other_companies:
            transport2 = providers[providers['id'] == company][['transport_type']].iloc[0]['transport_type']
            if(transport != transport2):
                return None
        return transport

In [18]:
tickets2['transport_type'] = tickets2.apply(lambda x: getTransportType(x['company'], x['other_companies']), axis = 1)

tickets2

id  company  o_station  d_station              departure_ts  \
0      6795025     8385         -1         -1 2017-10-13 14:00:00+00:00   
1      6795026        9         63       1044 2017-10-13 13:05:00+00:00   
2      6795027     8377       5905       6495 2017-10-13 13:27:00+00:00   
3      6795028     8377       5905       6495 2017-10-13 13:27:00+00:00   
4      6795029     8381       5905       6495 2017-10-13 21:46:00+00:00   
...        ...      ...        ...        ...                       ...   
74163  6869182       13        279      10729 2017-10-27 02:30:00+00:00   
74164  6869185        9        279        304 2017-10-27 07:00:00+00:00   
74165  6869187     8371      10642        304 2017-10-27 08:30:00+00:00   
74166  6869189       13        279        304 2017-10-27 13:25:00+00:00   
74167  6869191     8371      10642      10729 2017-10-27 14:45:00+00:00   

                     arrival_ts  price_in_cents                   search_ts  \
0     2017-10-13 20:10:00+00:00            4550  2017-10-01 00:13:31.327+00   
1     2017-10-14 06:55:00+00:00            1450  2017-10-01 00:13:35.773+00   
2     2017-10-14 21:24:00+00:00            7400  2017-10-01 00:13:40.212+00   
3     2017-10-14 11:02:00+00:00           13500  2017-10-01 00:13:40.213+00   
4     2017-10-14 19:32:00+00:00            7710  2017-10-01 00:13:40.213+00   
...                         ...             ...                         ...   
74163 2017-10-27 16:00:00+00:00            2590  2017-10-19 10:35:42.943+00   
74164 2017-10-27 13:30:00+00:00            2700  2017-10-19 10:35:42.943+00   
74165 2017-10-27 15:15:00+00:00            3000  2017-10-19 10:35:42.943+00   
74166 2017-10-27 21:10:00+00:00            2880  2017-10-19 10:35:42.943+00   
74167 2017-10-27 21:40:00+00:00            2000  2017-10-19 10:35:42.943+00   

           middle_stations other_companies  o_city  d_city       travel_ts  \
0                       -1              -1     611     542 0 days 06:10:00   
1                {149,418}            {13}     611     542 0 days 17:50:00   
2      {798,798,6794,6246}     {8377,8376}     611     542 1 days 07:57:00   
3      {798,798,6794,6246}     {8377,8376}     611     542 0 days 21:35:00   
4              {5983,5983}          {8380}     611     542 0 days 21:46:00   
...                    ...             ...     ...     ...             ...   
74163            {408,408}          {8371}     628     562 0 days 13:30:00   
74164          {1105,1105}             {9}     628     562 0 days 06:30:00   
74165          {1105,1105}             {9}     628     562 0 days 06:45:00   
74166            {863,863}            {13}     628     562 0 days 07:45:00   
74167            {863,863}          {8371}     628     562 0 days 06:55:00   

      transport_type  
0         carpooling  
1                bus  
2              train  
3              train  
4              train  
...              ...  
74163            bus  
74164            bus  
74165            bus  
74166            bus  
74167            bus  

[74168 rows x 14 columns]

In [19]:
tickets2.count()

id                 74168
company            74168
o_station          74168
d_station          74168
departure_ts       74168
arrival_ts         74168
price_in_cents     74168
search_ts          74168
middle_stations    74168
other_companies    74168
o_city             74168
d_city             74168
travel_ts          74168
transport_type     73885
dtype: int64

In [20]:
tickets2.dropna(inplace = True)
tickets2.count()

id                 73885
company            73885
o_station          73885
d_station          73885
departure_ts       73885
arrival_ts         73885
price_in_cents     73885
search_ts          73885
middle_stations    73885
other_companies    73885
o_city             73885
d_city             73885
travel_ts          73885
transport_type     73885
dtype: int64

Since it was verfy few lines, we could drop those is multiples transport type! 

We can the in the preview of the dataframe that every time a row hasn't 'middle_stations', it hasn't 'o_station' and 'd_station'. Let's check if it's correct (it'll make our computation of the distance a little bit easier).

In [21]:
def checkMiddle_o_d_stations(middle_stations, o_station, d_station):
    if(middle_stations != -1):
        if(o_station != -1 & d_station != -1):
            return True
        return False
    if(middle_stations == -1):
        if(o_station == -1 & d_station == -1):
            return True
        return False
    
testStations = tickets2.apply(lambda x: checkMiddle_o_d_stations(x['middle_stations'], x['o_station'], x['d_station']), axis = 1 )

all(testStations)

True

Nice! Now we know that if there is 'middle_stations', there are 'o_station' and 'd_station'. 

#### Distance

First, let's compute the distance. 

If we don't have any middle station we compute the distance between the **city of origin and the city of destination**. 
If there are middle stations, **we need to take them into account in the computation of the distance.**

Let's check how many trajects exists (including middle stations, o_station and d_station)

In [22]:
trajects_Test = tickets2.groupby(['o_city','d_city', 'middle_stations', 'o_station', 'd_station']).count()
trajects_Test.index

MultiIndex([(    5,   23,                     '{6797,1898}',   562,   269),
            (    6,  227,                       '{527,527}',  3653,   575),
            (    6,  227, '{6635,6635,470,470,10895,10895}',  3653,   575),
            (    6,  504,                                -1,    -1,    -1),
            (    6,  504,                     '{1044,1044}',    16,   763),
            (    6,  504,                       '{527,527}',    16,   334),
            (    6,  504,                       '{527,527}',    16,   763),
            (    6,  504,                       '{527,527}',    16, 10030),
            (    6,  504,                       '{527,527}',  3653,   334),
            (    6,  504,                     '{6242,6242}',  3653,   334),
            ...
            (11835,  246,                                -1,    -1,    -1),
            (11835,  485,                                -1,    -1,    -1),
            (11938,  126,                                -1,    -1,    -

Well, there are 6659 distances to compute (better than 74168)

Let's do this ! 

In [23]:
from math import sin, cos, sqrt, atan2, radians

# Method to compute distance from lat-lon coordinates 

def distanceCoord(lat1, lon1, lat2, lon2):
    R = 6373.0 # Approcimate radius of earth, in km
    
    lat1 = radians(lat1)
    lon1 = radians(lon1)
    
    lat2 = radians(lat2)
    lon2 = radians(lon2)
    
    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

def computeKmTrajects(o_city, d_city, middle_stations = -1, o_station = -1, d_station = 1):
    if(middle_stations != -1):
        dist = 0
        # We get the IDs of the middle stations in a set in order to get unique ids (as int) 
        middle_stations = set(map(int, middle_stations[1:-1].split(',')))
        
        # We compute the distance between every station
        previous_station = o_station
        for station in middle_stations:
            s1_lat_long = stations[stations['id'] == previous_station][['latitude','longitude']]    
            s2_lat_long = stations[stations['id'] == station][['latitude','longitude']]     

            dist += round(distanceCoord(s1_lat_long.latitude, s1_lat_long.longitude, s2_lat_long.latitude, s2_lat_long.longitude))
            previous_station = station
        
        s_lat_long = stations[stations['id'] == previous_station][['latitude','longitude']]  
        d_lat_long = stations[stations['id'] == d_station][['latitude','longitude']]
        
        # We compute the distance between the n-1 et n station ! Then return distance 
        dist += round(distanceCoord(s_lat_long.latitude, s_lat_long.longitude, d_lat_long.latitude, d_lat_long.longitude))
        return dist
        
    else:
        o_lat_long = cities[cities['id'] == o_city][['latitude','longitude']]
        d_lat_long = cities[cities['id'] == d_city][['latitude','longitude']]
        
        dist = round(distanceCoord(o_lat_long.latitude, o_lat_long.longitude, d_lat_long.latitude, d_lat_long.longitude))
        
        return dist

In [24]:
distances = dict()

for trajects in trajects_Test.index:
    distances[trajects] = computeKmTrajects(trajects[0], trajects[1], trajects[2], trajects[3], trajects[4])

In [25]:
def assignDistance(o_city, d_city, middle_stations = -1, o_station = -1, d_station = -1):
    return distances[(o_city,d_city,middle_stations,o_station,d_station)]

In [26]:
tickets2['travel_km'] = tickets2.apply(lambda x: assignDistance(x['o_city'], x['d_city'], x['middle_stations'], x['o_station'], x['d_station']), axis = 1)
tickets2

id  company  o_station  d_station              departure_ts  \
0      6795025     8385         -1         -1 2017-10-13 14:00:00+00:00   
1      6795026        9         63       1044 2017-10-13 13:05:00+00:00   
2      6795027     8377       5905       6495 2017-10-13 13:27:00+00:00   
3      6795028     8377       5905       6495 2017-10-13 13:27:00+00:00   
4      6795029     8381       5905       6495 2017-10-13 21:46:00+00:00   
...        ...      ...        ...        ...                       ...   
74163  6869182       13        279      10729 2017-10-27 02:30:00+00:00   
74164  6869185        9        279        304 2017-10-27 07:00:00+00:00   
74165  6869187     8371      10642        304 2017-10-27 08:30:00+00:00   
74166  6869189       13        279        304 2017-10-27 13:25:00+00:00   
74167  6869191     8371      10642      10729 2017-10-27 14:45:00+00:00   

                     arrival_ts  price_in_cents                   search_ts  \
0     2017-10-13 20:10:00+00:00            4550  2017-10-01 00:13:31.327+00   
1     2017-10-14 06:55:00+00:00            1450  2017-10-01 00:13:35.773+00   
2     2017-10-14 21:24:00+00:00            7400  2017-10-01 00:13:40.212+00   
3     2017-10-14 11:02:00+00:00           13500  2017-10-01 00:13:40.213+00   
4     2017-10-14 19:32:00+00:00            7710  2017-10-01 00:13:40.213+00   
...                         ...             ...                         ...   
74163 2017-10-27 16:00:00+00:00            2590  2017-10-19 10:35:42.943+00   
74164 2017-10-27 13:30:00+00:00            2700  2017-10-19 10:35:42.943+00   
74165 2017-10-27 15:15:00+00:00            3000  2017-10-19 10:35:42.943+00   
74166 2017-10-27 21:10:00+00:00            2880  2017-10-19 10:35:42.943+00   
74167 2017-10-27 21:40:00+00:00            2000  2017-10-19 10:35:42.943+00   

           middle_stations other_companies  o_city  d_city       travel_ts  \
0                       -1              -1     611     542 0 days 06:10:00   
1                {149,418}            {13}     611     542 0 days 17:50:00   
2      {798,798,6794,6246}     {8377,8376}     611     542 1 days 07:57:00   
3      {798,798,6794,6246}     {8377,8376}     611     542 0 days 21:35:00   
4              {5983,5983}          {8380}     611     542 0 days 21:46:00   
...                    ...             ...     ...     ...             ...   
74163            {408,408}          {8371}     628     562 0 days 13:30:00   
74164          {1105,1105}             {9}     628     562 0 days 06:30:00   
74165          {1105,1105}             {9}     628     562 0 days 06:45:00   
74166            {863,863}            {13}     628     562 0 days 07:45:00   
74167            {863,863}          {8371}     628     562 0 days 06:55:00   

      transport_type  travel_km  
0         carpooling        503  
1                bus        506  
2              train        753  
3              train        753  
4              train        630  
...              ...        ...  
74163            bus        346  
74164            bus        340  
74165            bus        344  
74166            bus        372  
74167            bus        379  

[73885 rows x 15 columns]

Now that we have the transport time and the transport type, we can move on !

The next thing to do is to make categories from travel_km

In [29]:
tickets2['travel_km_cat'] = pd.cut(x=tickets2['travel_km'], bins = [0, 200, 800, 2000, np.inf], labels = [1,2,3,4], include_lowest = True)
tickets2

id  company  o_station  d_station              departure_ts  \
0      6795025     8385         -1         -1 2017-10-13 14:00:00+00:00   
1      6795026        9         63       1044 2017-10-13 13:05:00+00:00   
2      6795027     8377       5905       6495 2017-10-13 13:27:00+00:00   
3      6795028     8377       5905       6495 2017-10-13 13:27:00+00:00   
4      6795029     8381       5905       6495 2017-10-13 21:46:00+00:00   
...        ...      ...        ...        ...                       ...   
74163  6869182       13        279      10729 2017-10-27 02:30:00+00:00   
74164  6869185        9        279        304 2017-10-27 07:00:00+00:00   
74165  6869187     8371      10642        304 2017-10-27 08:30:00+00:00   
74166  6869189       13        279        304 2017-10-27 13:25:00+00:00   
74167  6869191     8371      10642      10729 2017-10-27 14:45:00+00:00   

                     arrival_ts  price_in_cents                   search_ts  \
0     2017-10-13 20:10:00+00:00            4550  2017-10-01 00:13:31.327+00   
1     2017-10-14 06:55:00+00:00            1450  2017-10-01 00:13:35.773+00   
2     2017-10-14 21:24:00+00:00            7400  2017-10-01 00:13:40.212+00   
3     2017-10-14 11:02:00+00:00           13500  2017-10-01 00:13:40.213+00   
4     2017-10-14 19:32:00+00:00            7710  2017-10-01 00:13:40.213+00   
...                         ...             ...                         ...   
74163 2017-10-27 16:00:00+00:00            2590  2017-10-19 10:35:42.943+00   
74164 2017-10-27 13:30:00+00:00            2700  2017-10-19 10:35:42.943+00   
74165 2017-10-27 15:15:00+00:00            3000  2017-10-19 10:35:42.943+00   
74166 2017-10-27 21:10:00+00:00            2880  2017-10-19 10:35:42.943+00   
74167 2017-10-27 21:40:00+00:00            2000  2017-10-19 10:35:42.943+00   

           middle_stations other_companies  o_city  d_city       travel_ts  \
0                       -1              -1     611     542 0 days 06:10:00   
1                {149,418}            {13}     611     542 0 days 17:50:00   
2      {798,798,6794,6246}     {8377,8376}     611     542 1 days 07:57:00   
3      {798,798,6794,6246}     {8377,8376}     611     542 0 days 21:35:00   
4              {5983,5983}          {8380}     611     542 0 days 21:46:00   
...                    ...             ...     ...     ...             ...   
74163            {408,408}          {8371}     628     562 0 days 13:30:00   
74164          {1105,1105}             {9}     628     562 0 days 06:30:00   
74165          {1105,1105}             {9}     628     562 0 days 06:45:00   
74166            {863,863}            {13}     628     562 0 days 07:45:00   
74167            {863,863}          {8371}     628     562 0 days 06:55:00   

      transport_type  travel_km travel_km_cat  
0         carpooling        503             2  
1                bus        506             2  
2              train        753             2  
3              train        753             2  
4              train        630             2  
...              ...        ...           ...  
74163            bus        346             2  
74164            bus        340             2  
74165            bus        344             2  
74166            bus        372             2  
74167            bus        379             2  

[73885 rows x 16 columns]

Now that we have all the information we need (transport type and distance), we can proceed !

In [32]:
transport_travel = tickets2.groupby(['transport_type','travel_km_cat'])[['price_in_cents','travel_ts']].agg(['min','max', 'sum', 'count'])
transport_travel

price_in_cents                               \
                                        min      max          sum  count   
transport_type travel_km_cat                                               
bus            1                      850.0   7800.0    1045026.0    547   
               2                      980.0  12800.0   38956848.0  11679   
               3                     2290.0  22900.0    9110804.0   1379   
               4                     5000.0  16890.0     202960.0     24   
carpooling     1                      300.0  12850.0   12340950.0  10330   
               2                      850.0  13800.0  100063800.0  30968   
               3                     4496.0  16150.0    1233596.0    143   
               4                        NaN      NaN          NaN      0   
train          1                      490.0   8120.0    6099990.0   2059   
               2                     1400.0  33900.0  138147590.0  15607   
               3                     1790.0  38550.0   15816330.0   1132   
               4                     9100.0  22030.0     268590.0     17   

                                   travel_ts                   \
                                         min              max   
transport_type travel_km_cat                                    
bus            1             0 days 01:28:00 13 days 05:45:00   
               2             0 days 04:00:00 20 days 12:51:00   
               3             0 days 11:50:00 14 days 17:00:00   
               4             0 days 22:20:00  8 days 02:10:00   
carpooling     1             0 days 00:20:00  0 days 17:20:00   
               2             0 days 01:20:00  0 days 19:50:00   
               3             0 days 07:50:00  1 days 05:10:00   
               4                         NaT              NaT   
train          1             0 days 00:39:00  1 days 10:04:00   
               2             0 days 01:08:00  2 days 00:27:00   
               3             0 days 02:32:00  1 days 18:30:00   
               4             0 days 07:04:00  1 days 00:09:00   

                                                        
                                            sum  count  
transport_type travel_km_cat                            
bus            1              177 days 22:13:00    547  
               2             7192 days 22:56:00  11679  
               3             1520 days 02:01:00   1379  
               4               38 days 04:55:00     24  
carpooling     1              851 days 03:20:00  10330  
               2             6168 days 16:40:00  30968  
               3               78 days 01:20:00    143  
               4                            NaT      0  
train          1              368 days 22:56:00   2059  
               2             4785 days 06:05:00  15607  
               3              572 days 03:04:00   1132  
               4               11 days 09:32:00     17